File: tests/common/src/cpp/coherence/tests/ProxyControlInvocable.cpp
/*
 * Copyright (c) 2000, 2020, Oracle and/or its affiliates.
 *
 * Licensed under the Universal Permissive License v 1.0 as shown at
 * http://oss.oracle.com/licenses/upl.
 */
#include "coherence/tests/ProxyControlInvocable.hpp"

#include "coherence/io/pof/PofReader.hpp"
#include "coherence/io/pof/PofWriter.hpp"
#include "coherence/io/pof/SystemPofContext.hpp"

COH_OPEN_NAMESPACE2(coherence,tests)

COH_REGISTER_PORTABLE_CLASS(1600, ProxyControlInvocable);

using coherence::io::pof::PofReader;
using coherence::io::pof::PofWriter;

// ----- constructors -------------------------------------------------------

ProxyControlInvocable::ProxyControlInvocable()
        : f_vsService(self(), NULL),
          m_nOperation(0)
    {
    }

ProxyControlInvocable::ProxyControlInvocable(String::View vsService, int32_t nOperation)
        : f_vsService(self(), vsService),
          m_nOperation(nOperation)
    {
    }

// ----- PortableObject interface -------------------------------------------

void ProxyControlInvocable::readExternal(PofReader::Handle hIn)
    {
    initialize(f_vsService, hIn->readString(0));
    m_nOperation = hIn->readInt32(1);
    }

void ProxyControlInvocable::writeExternal(PofWriter::Handle hOut) const
    {
    hOut->writeString(0, f_vsService);
    hOut->writeInt32(1, m_nOperation);
    }

COH_CLOSE_NAMESPACE2

File: tests/common/src/java/coherence/tests/ProxyControlInvocable.java
/*
 * Copyright (c) 2000, 2020, Oracle and/or its affiliates.
 *
 * Licensed under the Universal Permissive License v 1.0 as shown at
 * http://oss.oracle.com/licenses/upl.
 */
package coherence.tests;

import com.tangosol.io.pof.PortableObject;
import com.tangosol.io.pof.PofReader;
import com.tangosol.io.pof.PofWriter;

import com.tangosol.net.CacheFactory;
import com.tangosol.net.AbstractInvocable;

import java.io.IOException;

/**
* Invocable implementation that controls a proxy service.
*
* @author par  2012.12.12
*/
public class ProxyControlInvocable
        extends AbstractInvocable
        implements PortableObject
    {
    // ----- constructors ---------------------------------------------------

    /**
    * Default constructor.
    */
    public ProxyControlInvocable()
        {
        }

    // ----- Invocable interface --------------------------------------------

    /**
    * {@inheritDoc}
    */
    public void run()
        {
        String sService = getProxyServiceName();

        if ((m_nOperation == STOP_OPERATION || m_nOperation == RESTART_OPERATION) &&
            CacheFactory.getCluster().getService(sService) != null)
            {
            // go through the "front door" to get the SafeService
            CacheFactory.getService(sService).shutdown();
            }

        if (m_nOperation == START_OPERATION || m_nOperation == RESTART_OPERATION)
            {
            CacheFactory.getService(sService).start();
            }
        }

    // ----- PortableObject interface ---------------------------------------

    /**
    * {@inheritDoc}
    */
    public void readExternal(PofReader in)
            throws IOException
        {
        setProxyServiceName(in.readString(0));
        m_nOperation = in.readInt(1);
        }

    /**
    * {@inheritDoc}
    */
    public void writeExternal(PofWriter out)
            throws IOException
        {
        out.writeString(0, getProxyServiceName());
        out.writeInt(1, m_nOperation);
        }

    // ----- accessors ------------------------------------------------------

    /**
    * Return the proxy service to stop.
    *
    * @return the proxy service name
    */
    public String getProxyServiceName()
        {
        return m_proxyServiceName;
        }

    /**
    * Configure the proxy service to stop.
    *
    * @param name  the proxy service to stop
    */
    public void setProxyServiceName(String name)
        {
        m_proxyServiceName = name;
        }

    // ----- constants ------------------------------------------------------
    
    /**
     * Start the service.
     */
    static public final int START_OPERATION = 1;

    /**
     * Stop the service.
     */
    static public final int STOP_OPERATION = 2;

    /**
     * Restart the service.
     */
    static public final int RESTART_OPERATION = 3;

    // ----- data members ---------------------------------------------------

    /**
    * The service name to stop.
    */
    protected String m_proxyServiceName;

    /**
    * Operation
    */
    protected int m_nOperation;
    }
